types/types.cc

```diff
@@ -2851,6 +3044,25 @@ struct from_string_visitor {
         }
         return concat_fields(fields, field_len);
     }
+    bytes operator()(const vector_type_impl& t) {
+        std::vector<std::string_view> field_strings = split_field_strings(s, ',');
+        if (field_strings.size() > t.get_dimension()) {
+            throw marshal_exception(
+                    format("Invalid vector literal: too many elements. Type {} expects {:d} but got {:d}",
+                            t.as_cql3_type(), t.get_dimension(), field_strings.size()));
+        }
+        std::vector<bytes> fields(field_strings.size());
+        std::vector<int32_t> field_len(field_strings.size());
+        for (std::size_t i = 0; i < field_strings.size(); ++i) {
+            field_strings[i] = trim(field_strings[i]);
+            fields[i] = t.get_elements_type()->from_string(field_strings[i]);
+            field_len[i] = fields[i].size();
+        }
+        if (t.get_elements_type()->is_native()) {
+            return concat_fields(fields);
+        }
+        return concat_fields(fields, field_len);
+    }
```

Comment:
It is tested [here](https://github.com/zpp-2024-vector-search/scylladb/blob/83604d7f9867771776c805fe82377439552141e3/test/boost/types_test.cc#L612C5-L617C56) similar to the tuple type.

Reply:
Can we change this test to be a cqlpy test, so that you can run it against Cassandra as well?

Reply:
The test has been added via df2a84d4570978f74e4f5464ea5187e4f1590aa1.

Although Cassandra seems not to support such way of insertion, it works fine with our implementation.
The syntax for this was based on [to_string ](https://github.com/zpp-2024-vector-search/scylladb/blob/df2a84d4570978f74e4f5464ea5187e4f1590aa1/types/types.cc#L3094C1-L3156C3) function which acts the same for list, set and vector, and outputs their elements in a string split by commas.

Nevertheless I'm not sure now if this functionality should be added, if it is not supported the same way in Cassandra
(seems like it only works with tuples, I was not able to deduce the correct syntax for collections and vectors).
Furthermore collections don't implement it as well [(marked as FIXME)](https://github.com/zpp-2024-vector-search/scylladb/blob/df2a84d4570978f74e4f5464ea5187e4f1590aa1/types/types.cc#L3046C5-L3050C6).

Could you express your opinion on this case?

Reply:
> Although Cassandra seems not to support such way of insertion, it works fine with our implementation.

If Cassandra does not support this syntax then there is a risk that it will in the future, and will decide to do so in an incompatible way to Scylla's. In that case I'd prefer we didn't introduce support for it.

> [(marked as FIXME)](https://github.com/zpp-2024-vector-search/scylladb/blob/df2a84d4570978f74e4f5464ea5187e4f1590aa1/types/types.cc#L3046C5-L3050C6)

I don't think we should outright abort the node if we try it (by the way, can this abort be triggered now with collections?), but we should return a similar error to what Cassandra returns.

Reply:
> I don't think we should outright abort the node if we try it (by the way, can this abort be triggered now with collections?), but we should return a similar error to what Cassandra returns.

As far as I've checked, trying to insert into a list or set column using string results in InvalidRequest exception **from the driver**, so the abort() function is not called as it is adressed by this exception before.

Although driver seems not to support vectors in such case, and allows the execution of from_string_visitor (leading to abort() function if we removed the implementation).

Furthermore, in Cassandra it works OK with vectors, the same exception as for lists and sets.

I'm still not sure how should we deal with this problem, further ideas highly appreciated.